builtin/providers/cloudstack/provider.go

```diff
@@ -72,10 +88,42 @@ func Provider() terraform.ResourceProvider {
 }
 
 func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	apiURL := d.Get("api_url").(string)
+	apiKey := d.Get("api_key").(string)
+	secretKey := d.Get("secret_key").(string)
+
+	if configFile := d.Get("config").(string); configFile != "" {
+		config, err := ini.Load(configFile)
+		if err != nil {
+			return config, err
+		}
+
+		section, err := config.GetSection(d.Get("profile").(string))
+		if err != nil {
+			return config, err
+		}
+
+		if apiURL == "" {
+			apiURL = section.Key("url").String()
+		}
+
+		if apiKey == "" {
+			apiKey = section.Key("apikey").String()
+		}
+
+		if secretKey == "" {
+			secretKey = section.Key("secretkey").String()
+		}
```

Comment:
Why do you only assign the config values if the variables are empty? Either all the values come from the config file, or they all come from the individual `api_url`, `api_key` and `secret_key`  parameters. To prevent unexpected behaviour they should never be mixed...